### PR TITLE
Fix BackendType repr in doc

### LINKED ISSUE #29905: Argument missing and ordering mismatch at `torch.distributed.rpc.init_model_parallel` doc
## 📚 Documentation

Per `torch.distributed.rpc.init_model_parallel` function on RPC document page in master branch [here](https://pytorch.org/docs/master/rpc.html), I think there are some points to revise or polish:

1. the argument `worker_name_to_id` is missing in doc.
2. The order of arguments in doc does not match the function signature, which would cause additional ambiguity or inconvenience for readers.
3. I think the repr for argument `backend` is too long to read, which may contain a lot of unneeded information for readers. It is arguable but it might be better to give the class a concise repr.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528

### PR DESCRIPTION
**Pull request by mrshenli**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30259 Fix two links in RPC API doc
* **#30243 Fix BackendType repr in doc**
* #30240 Fix RRef design doc warning
* #30218 Add link to RRef protocol in RPC doc

Before this commit, rpc docs shows init_rpc as the following:

```python
torch.distributed.rpc.init_rpc(
   name,
   backend=<BackendType.PROCESS_GROUP: BackendValue(
     construct_rpc_agent_options_handler=<function _process_group_construct_rpc_agent_options_handler>,
     init_backend_handler=<function _process_group_init_backend_handler>)>,
   init_method=None,
   rank=-1,
   world_size=None,
   rpc_agent_options=None
)
```

It unnecessarily leaks implementation details. This commit adds a
__repr__ function to BackendType Enum class to address this problem.

closes #29905

Differential Revision: [D18641559](https://our.internmc.facebook.com/intern/diff/D18641559)

### CODE REVIEW

**Comment by mrshenli** [comment]
Screen Shot 2019-11-21 at 1 54 45 PM

**Review by rohan-varma — APPROVED** [review]
Looks good. Does `BackendType` show up in docs? Would we need it to show up so users know about `BackendType.ProcessGroup`?

**Comment by mrshenli** [comment]
> Looks good. Does BackendType show up in docs? Would we need it to show up so users know about BackendType.ProcessGroup?

Good point, it does not yet, and we should have it. Let me try in a followup PR.

File: torch/distributed/rpc/backend_registry.py [hunk 1]
```diff
@@ -14,8 +14,17 @@
...
+def _backend_type_repr(self):
+    return "BackendType." + self.name
+
+
 # Create an enum type, `BackendType`, with empty members.
-BackendType = enum.Enum(value="BackendType", names={})
+BackendType = enum.Enum(
+    value="BackendType",
+    names={},
+    __repr__=_backend_type_repr
```
Comment (xush6528) [inline torch/distributed/rpc/backend_registry.py]: How does this work.

It looks to me extra keyword arg will be dropped by this asterisk.

https://github.com/python/cpython/blob/293dd23477eef6e7c1b1e26b5bb2c1e0d79ac3c2/Lib/enum.py#L278

Comment (mrshenli) [reply torch/distributed/rpc/backend_registry.py]: Ah, yes you are right! I was doing `BackendType .__repr__ = _backend_type_repr` when generating the figure and then switched to this afterward. It was showing the same result, but I should have `make clean` the docs, after that I started to see errors. Let me update.

**Review by xush6528 — APPROVED** [review]
I feel certain that this works!

**Comment by pritamdamania87** [comment]
@mrshenli I see a couple of issues with the docs:

1. `init_rpc` used to be a link to the method, now the link is lost.
2. The following line doesn't seem right: "By default, this will also initialize the ProcessGroup (torch.distributed.Backend()) backend for RPC communication". We shouldn't link to torch.distributed.Backend (since this backend refers to gloo,nccl and mpi) for ProcessGroup. We should link to it when we specify that we use gloo for communication. If we do want a link for ProcessGroup, maybe we can link to https://pytorch.org/docs/master/distributed.html#torch.distributed.init_process_group for now?

**Comment by mrshenli** [comment]
@pritamdamania87 

> 1. init_rpc used to be a link to the method, now the link is lost.

Where is that link?

**Comment by mrshenli** [comment]
> The following line doesn't seem right: "By default, this will also initialize the ProcessGroup (torch.distributed.Backend()) backend for RPC communication". We shouldn't link to torch.distributed.Backend (since this backend refers to gloo,nccl and mpi) for ProcessGroup. We should link to it when we specify that we use gloo for communication. If we do want a link for ProcessGroup, maybe we can link to https://pytorch.org/docs/master/distributed.html#torch.distributed.init_process_group for now?

I agree. This is irrelevant to this PR though, add fix in a followup PR

**Comment by mrshenli** [comment]
Addressed @pritamdamania87's comments in #30259

**Pull request closed without merging**